### ISSUE TITLE
`tf.nn.selu` and `tf.nn.elu` return 0 for small negative inputs (catastrophic cancellation in `exp(x) - 1`)

### ISSUE THREAD

**Issue by wuyii8941** [tensorflow/tensorflow#116943]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

TensorFlow 2.22.0-dev20260421 (nightly)

### Custom code

Yes

### Current behavior?

# `tf.nn.selu` and `tf.nn.elu` return 0 for small negative inputs (catastrophic cancellation in `exp(x) - 1`)

## Summary

`tf.nn.selu` and `tf.nn.elu` compute `exp(x) - 1` via separate `exp` and subtraction. For small negative x, `exp(x)` rounds to `1.0` in float32, producing `exp(x) - 1 = 0` (catastrophic cancellation). The result is 100% wrong for |x| ≤ 1e-8, and already 1.3% wrong at x = -1e-6.

XLA (`jit_compile=True`) is not affected — it uses `expm1` internally and returns correct results.

## Scale of the problem

SELU(x) for x < 0 is `scale × alpha × (exp(x) - 1)`. For small negative x, the correct value is approximately `1.7581 × x`.

| x | SELU expected | SELU eager | Relative error |
|---|----------|---------------|----------------|
| -1e-04 | -1.758e-4 | -1.758e-4 | 0.02% |
| -1e-06 | -1.758e-6 | -1.781e-6 | **1.33%** |
| -1e-07 | -1.758e-7 | -2.096e-7 | **19.21%** |
| -1e-08 | -1.758e-8 | **0.0** | **100%** |
| -1e-10 | -1.758e-10 | **0.0** | **100%** |

ELU has the same pattern (with alpha = 1):

| x | ELU expected | ELU eager | Relative error |
|---|----------|---------------|----------------|
| -1e-06 | -1.000e-6 | -1.013e-6 | **1.33%** |
| -1e-07 | -1.000e-7 | -1.192e-7 | **19.21%** |
| -1e-08 | -1.000e-8 | **0.0** | **100%** |

### Gradient is also affected (SELU)

Because the forward pass returns 0 for small negative x, the backward pass takes the wrong branch (`scale` instead of `scale × alpha × exp(x)`), producing ~40% gradient error:

```
x = -1e-07: selu_grad = 1.7581 (correct)
x = -1e-08: selu_grad = 1.0507 (wrong — should be 1.7581, ~40% error)
```

## Root cause

```
exp(-1e-8)     = 1.0       (in float32 — rounds to 1)
exp(-1e-8) - 1 = 0.0       (catastrophic cancellation)
expm1(-1e-8)   = -1e-8     (correct, no cancellation)
```

The fix is to replace `exp(x) - 1` with `expm1(x)` in the Eigen SELU/ELU kernels. XLA already does this, which is why `jit_compile=True` produces correct results.

## Environment

- TensorFlow 2.22.0-dev20260421 (nightly)
- CPU and GPU eager kernels both affected
- XLA is NOT affected

### Standalone code to reproduce the issue

```shell
## Reproduction


import tensorflow as tf

x = tf.constant([[-1e-10]], dtype=tf.float32)

print(tf.nn.selu(x).numpy())   # [[0.]]  — should be [[-1.758e-10]]
print(tf.nn.elu(x).numpy())    # [[0.]]  — should be [[-1e-10]]

# XLA gives the correct result:
print(tf.function(tf.nn.selu, jit_compile=True)(x).numpy())  # [[-1.758e-10]]
print(tf.function(tf.nn.elu, jit_compile=True)(x).numpy())   # [[-1e-10]]
```

### Relevant log output

```shell

```

**Comment by Mahajan-Sachin** [comment]
Hi, I'd like to submit a fix for this issue.

The root cause is clear: `exp(x) - 1` suffers from catastrophic cancellation in float32 when `|x|` is small. I've prepared a PR that replaces `exp(x) - 1` with `expm1(x)` in three locations:

1. `Elu` functor in `relu_op_functor.h`  
2. `Selu` functor in `relu_op_functor.h`  
3. Fused `Elu` in `fused_eigen_output_kernels.h`

Eigen's `scalar_expm1_op` is already used by TensorFlow's `Expm1` op, so this is a well-tested code path. The fix aligns eager kernel behavior with XLA, which already uses `expm1` internally.

**Comment by Mahajan-Sachin** [comment]
I've submitted a fix in PR #116947.

**Before fix** — `exp(x) - 1` produces 0.0 for small negative x (100% error):

<img width="940" height="899" alt="Image" src="https://github.com/user-attachments/assets/609cf6fa-86ae-4a09-9d5c-8241e7a008c1" />

**After fix** — `expm1(x)` returns correct values (0.00% error):

<img width="774" height="626" alt="Image" src="https://github.com/user-attachments/assets/bfedd32a-1dbd-4f8b-810b-e8ae33bc58ce" />

The fix replaces `exp(x) - 1` with `expm1(x)` in the Eigen ELU/SELU kernels, matching XLA behavior.

**Comment by Kayyuri** [comment]
Hi @wuyii8941, Thank you raising this issue.
I was able to replicate the issue reported [here](https://colab.sandbox.google.com/gist/Kayyuri/912c58e7c51848d5a7fc4bed74db088b/116943.ipynb). A [pull request](https://github.com/tensorflow/tensorflow/pull/116947) has been submitted to fix this issue, and it will be resolved once merged.

Thanks to @Mahajan-Sachin for raising the PR to address this issue. We truly appreciate your initiative and the effort in contributing the fix.

**Comment by Mahajan-Sachin** [comment]
Thank you for the kind words and for reviewing the PR! Happy to contribute. Let me know if any changes are needed.
